packages/playwright/src/runner/watchMode.ts

```diff
@@ -15,130 +15,127 @@
-  // Reset the settings that don't apply to watch.
-  config.cliPassWithNoTests = true;
-  for (const p of config.projects)
-    p.project.retries = 0;
-
-  // Perform global setup.
-  const testRun = new TestRun(config);
-  const taskRunner = createTaskRunnerForWatchSetup(config, [new ListReporter()]);
-  taskRunner.reporter.onConfigure(config.config);
-  const { status, cleanup: globalCleanup } = await taskRunner.runDeferCleanup(testRun, 0);
-  if (status !== 'passed')
-    await globalCleanup();
-  await taskRunner.reporter.onEnd({ status });
-  await taskRunner.reporter.onExit();
-  if (status !== 'passed')
-    return status;
-
-  // Prepare projects that will be watched, set up watcher.
-  const failedTestIdCollector = new Set<string>();
-  const originalWorkers = config.config.workers;
-  const fsWatcher = new FSWatcher();
-  await fsWatcher.update(config);
-
-  let lastRun: { type: 'changed' | 'regular' | 'failed', failedTestIds?: Set<string>, dirtyTestFiles?: Map<FullProjectInternal, Set<string>> } = { type: 'regular' };
+interface WatchModeOptions {
+  files?: string[];
+  projects?: string[];
+  grep?: string;
+}
+
+export async function runWatchModeLoop(configLocation: ConfigLocation, initialOptions: WatchModeOptions): Promise<FullResult['status']> {
+  const options: WatchModeOptions = { ...initialOptions };
+
+  const testServerDispatcher = new TestServerDispatcher(configLocation);
+  const transport = new InMemoryTransport(
+      async data => {
+        const { id, method, params } = JSON.parse(data);
+        try {
+          const result = await testServerDispatcher.transport.dispatch(method, params);
+          transport.emit('message', JSON.stringify({ id, result }));
+        } catch (e) {
+          transport.emit('message', JSON.stringify({ id, error: String(e) }));
+        }
+      }
+  );
+  testServerDispatcher.transport.sendEvent = (method, params) => {
+    transport.emit('message', JSON.stringify({ method, params }));
+  };
+  const testServerConnection = new TestServerConnection(transport);
+  transport.emit('open');
+
+  const teleSuiteUpdater = new TeleSuiteUpdater({ pathSeparator: path.sep, onUpdate() { } });
+
+  const dirtyTestIds = new Set<string>();
+  const onDirtyTests: { resolve?(): void } = {};
+
+  let queue = Promise.resolve();
+
+  testServerConnection.onTestFilesChanged(({ testFiles: changedFiles }) => {
+    if (changedFiles.length === 0)
+      return;
+
+    queue = queue.then(async () => {
+      const { report } = await testServerConnection.listTests({ locations: options.files, projects: options.projects, grep: options.grep });
+      teleSuiteUpdater.processListReport(report);
+
+      for (const test of teleSuiteUpdater.rootSuite!.allTests()) {
+        if (changedFiles.includes(test.location.file))
+          dirtyTestIds.add(test.id);
+      }
+
+      // listTests can take longer than the debouncing interval of onTestFilesChanged,
```

Comment:
I'd instead add `changedFiles` from multiple `onTestFilesChanged` into a single set, while a single `listTests()` is in progress, and then run lines 107-110 against the combined set and trigger `onDirtyTests` synchronously. This will make it a single async hop that's easier to reason about.